### PR TITLE
rgw: Remove unnecessary null check from valid_s3_bucket_name

### PR DESCRIPTION
**Pull request by vedanshbhartia**
## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by soumyakoduri** [comment]
> @soumyakoduri @mattbenjamin Could you please review this? When checking whether an s3 bucket name contains a dash or a dot adjacent to a dot, this change accounts for whether the current dot is the first or the last character. Previously, even if the dot was the first character, we would try to access the previous byte and validate it, and there was a similar issue with the last character. cc @yuvalif


Bucket name/label must start and end with a lowercase letter or a number.. This is already checked at line:853, line:862 before this `for loop`.

File: src/rgw/rgw_rest_s3.h [hunk 1]
```diff
@@ -862,39 +862,41 @@ inline int valid_s3_bucket_name(const std::string& name, bool relaxed=false)
...
-	continue;
+    if(c == '.') {
+      if(relaxed) {
+        continue;
       }
-    }
 
+      // name cannot contain a dot or a dash adjacent to a dot
+      if(
+        (itr == name.begin() || (*(itr-1) != '.' && *(itr-1) != '-')) &&
```
Comment (yuvalif) [inline src/rgw/rgw_rest_s3.h:893]: in the original code we check that `p != '-'` we do not check for not being a dot

**Comment by yuvalif** [comment]
> > @soumyakoduri @mattbenjamin Could you please review this? When checking whether an s3 bucket name contains a dash or a dot adjacent to a dot, this change accounts for whether the current dot is the first or the last character. Previously, even if the dot was the first character, we would try to access the previous byte and validate it, and there was a similar issue with the last character. cc @yuvalif
> 
> Bucket name/label must start and end with a lowercase letter or a number.. This is already checked at line:853, line:862 before this `for loop`.

but what if the last charecter is a dot, and `relaxed = true`. in this case we won't return ann error in line 863 and would try to access after the end inside the loop

**Comment by soumyakoduri** [comment]
> > but what if the last charecter is a dot, and `relaxed = true`. in this case we won't return ann error in line 863 and would try to access after the end inside the loop
> 
> it would then reach else {} loop at line:893 (in the original code)

[Sorry by mistake, I edited your original comment itself instead of replying..]  .. It may be ok to reformat the code here but I think there is no issue as mentioned in the description in the existing code.

**Comment by yuvalif** [comment]
> > > but what if the last charecter is a dot, and `relaxed = true`. in this case we won't return ann error in line 863 and would try to access after the end inside the loop
> > 
> > 
> > it would then reach else {} loop at line:893 (in the original code)
> 
> [Sorry by mistake, I edited your original comment itself instead of replying..] .. It may be ok to reformat the code here but I think there is no issue as mentioned in the description in the existing code.

@soumyakoduri can you please re-review?

**Comment by soumyakoduri** [comment]
> > > > but what if the last charecter is a dot, and `relaxed = true`. in this case we won't return ann error in line 863 and would try to access after the end inside the loop
> > > 
> > > 
> > > it would then reach else {} loop at line:893 (in the original code)
> > 
> > 
> > [Sorry by mistake, I edited your original comment itself instead of replying..] .. It may be ok to reformat the code here but I think there is no issue as mentioned in the description in the existing code.
> 
> @soumyakoduri can you please re-review?

As mentioned in the previous comments, there is no issue with the original code. IMO this coverity can be closed as false report. If there is any other issue being addressed, kindly update the description.

**Comment by yuvalif** [comment]
> > > > > but what if the last charecter is a dot, and `relaxed = true`. in this case we won't return ann error in line 863 and would try to access after the end inside the loop
> > > > 
> > > > 
> > > > it would then reach else {} loop at line:893 (in the original code)
> > > 
> > > 
> > > [Sorry by mistake, I edited your original comment itself instead of replying..] .. It may be ok to reformat the code here but I think there is no issue as mentioned in the description in the existing code.
> > 
> > 
> > @soumyakoduri can you please re-review?
> 
> As mentioned in the previous comments, there is no issue with the original code. IMO this coverity can be closed as false report. If there is any other issue being addressed, kindly update the description.

@soumyakoduri regarding the existign code, please consider the following case:
lets assume that the name ends with a dot and `relaxed = false`
this means we would not return in line 863.
would enter the "if" statements in lines 884 and 885, and would dereference a value after the last one.
i think this is working because this is the null terminator, but I don't think this is the original intent of the code

**Comment by soumyakoduri** [comment]
> > > > > > but what if the last charecter is a dot, and `relaxed = true`. in this case we won't return ann error in line 863 and would try to access after the end inside the loop
> > > > > 
> > > > > 
> > > > > it would then reach else {} loop at line:893 (in the original code)
> > > > 
> > > > 
> > > > [Sorry by mistake, I edited your original comment itself instead of replying..] .. It may be ok to reformat the code here but I think there is no issue as mentioned in the description in the existing code.
> > > 
> > > 
> > > @soumyakoduri can you please re-review?
> > 
> > 
> > As mentioned in the previous comments, there is no issue with the original code. IMO this coverity can be closed as false report. If there is any other issue being addressed, kindly update the description.
> 
> @soumyakoduri regarding the existign code, please consider the following case: lets assume that the name ends with a dot and `relaxed = false` this means we would not return in line 863. would enter the "if" statements in lines 884 and 885, and would dereference a value after the last one. i think this is working because this is the null terminator, but I don't think this is the original intent of the code

Thanks Yuval. But as mentioned earlier, this case cannot arise as the bucket name/label **must start and end** with a lowercase letter or a number.. This is already checked at line:853, line:862 before this `for loop` in the existing code.
(https://github.com/ceph/ceph/blob/main/src/rgw/rgw_rest_s3.h#L852-L863)

@vedanshbhartia please consider the checks outside this `for loop` too . If you still see any issue in the code, please verify and ensure all the constraints listed in https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html & https://docs.ceph.com/en/latest/radosgw/s3/bucketops/ are met for any changes you may make in this routine. Thanks!

**Comment by yuvalif** [comment]
> > > > > > > but what if the last charecter is a dot, and `relaxed = true`. in this case we won't return ann error in line 863 and would try to access after the end inside the loop
> > > > > > 
> > > > > > 
> > > > > > it would then reach else {} loop at line:893 (in the original code)
> > > > > 
> > > > > 
> > > > > [Sorry by mistake, I edited your original comment itself instead of replying..] .. It may be ok to reformat the code here but I think there is no issue as mentioned in the description in the existing code.
> > > > 
> > > > 
> > > > @soumyakoduri can you please re-review?
> > > 
> > > 
> > > As mentioned in the previous comments, there is no issue with the original code. IMO this coverity can be closed as false report. If there is any other issue being addressed, kindly update the description.
> > 
> > 
> > @soumyakoduri regarding the existign code, please consider the following case: lets assume that the name ends with a dot and `relaxed = false` this means we would not return in line 863. would enter the "if" statements in lines 884 and 885, and would dereference a value after the last one. i think this is working because this is the null terminator, but I don't think this is the original intent of the code
> 
> Thanks Yuval. But as mentioned earlier, this case cannot arise as the bucket name/label **must start and end** with a lowercase letter or a number.. This is already checked at line:853, line:862 before this `for loop` in the existing code. (https://github.com/ceph/ceph/blob/main/src/rgw/rgw_rest_s3.h#L852-L863)
> 
> @vedanshbhartia please consider the checks outside this `for loop` too . If you still see any issue in the code, please verify and ensure all the constraints listed in https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html & https://docs.ceph.com/en/latest/radosgw/s3/bucketops/ are met for any changes you may make in this routine. Thanks!

ok. got it. so, if `relaxed == false` we would return before the loop, and if `relaxed == true` we would not check the prev/next values

**Review by yuvalif — APPROVED** [review]

**Review by soumyakoduri — APPROVED** [review]

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
jenkins test api

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
> teuthology results: http://pulpito.ceph.com/yuvalif-2023-10-08_18:12:17-rgw-wip-yuval-test-gsoc-distro-default-smithi/

same as baseline: https://pulpito.ceph.com/yuriw-2023-10-07_14:02:30-rgw-main-distro-default-smithi/

**Pull request merged**